### PR TITLE
Use ratios

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
Use Ratios in Loan and callSpread contracts.

The loan has support for backward compatibility, since the MMR was visible in the terms as a Percent.

#documentation-branch: ratios

### CODE REVIEW

File: packages/zoe/src/contracts/loan/borrow.js [hunk 1]
```diff
@@ -15,14 +15,22 @@ import { makeAddCollateralInvitation } from './addCollateral';
 /** @type {MakeBorrowInvitation} */
 export const makeBorrowInvitation = (zcf, config) => {
   const {
-    mmr, // Maintenance Margin Requirement, in percent
+    mmr, // Maintenance Margin Requirement, as a Percent (deprecated)
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/borrow.js]: Is there a reason not to change this now? I think nothing depends on the loan using a mmr percent. Seems like more work and more confusing to keep as a percent

File: packages/zoe/src/contracts/loan/borrow.js [hunk 2]
```diff
@@ -48,16 +56,16 @@ export const makeBorrowInvitation = (zcf, config) => {
...
+    // formula: assert collateralValue*100 >= loanWanted*mmrRatio
 
     // Calculate approximate value just for the error message if needed
-    const approxForMsg = mmr.scale(loanWanted);
+    const approxForMsg = multiplyBy(loanWanted, mmrRatio);
 
     // Assert the required collateral was escrowed.
     assert(
       loanMath.isGTE(
         loanMath.make(collateralPriceInLoanBrand.value),
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/borrow.js]: This didn't change in this PR, but I just realized this line doesn't make any sense. we are remaking an amount from an amount.

File: packages/zoe/src/contracts/loan/borrow.js [hunk 3]
```diff
@@ -48,16 +56,16 @@ export const makeBorrowInvitation = (zcf, config) => {
 
     const collateralPriceInLoanBrand = quoteAmount.value[0].amountOut;
 
-    // formula: assert collateralValue*100 >= loanWanted*mmr
+    // formula: assert collateralValue*100 >= loanWanted*mmrRatio
 
     // Calculate approximate value just for the error message if needed
-    const approxForMsg = mmr.scale(loanWanted);
+    const approxForMsg = multiplyBy(loanWanted, mmrRatio);
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/borrow.js]: This isn't approximate anymore, so we can just call this `requiredMargin` and reuse it below.

File: packages/zoe/src/contracts/loan/borrow.js [hunk 4]
```diff
@@ -15,14 +15,22 @@ import { makeAddCollateralInvitation } from './addCollateral';
...
   } = config;
 
+  // TODO(hibbert) drop mmr as Percent before Beta
+  let {
+    mmrRatio, // Maintenance Margin Requirement, as a ratio
+  } = config;
+  if (!mmrRatio) {
+    mmrRatio = mmr.makeRatio();
+  }
+
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/borrow.js]: borrow is internal to the loan contract, so it doesn't make sense to pass around mmr - we should have already changed it to a Ratio at this point, after hitting index. However, I think it's best to just drop the percent and use ratios directly anyways.

File: packages/zoe/src/contracts/loan/scheduleLiquidation.js [hunk 5]
```diff
@@ -17,12 +18,20 @@ export const scheduleLiquidation = (zcf, configWithBorrower) => {
...
   } = configWithBorrower;
 
+  // TODO(hibbert) drop mmr as Percent before Beta
+  let {
+    mmrRatio, // Maintenance Margin Requirement, as a ratio
+  } = configWithBorrower;
+  if (!mmrRatio) {
+    mmrRatio = mmr.makeRatio();
+  }
+
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/scheduleLiquidation.js]: This is internal, so it doesn't make sense to pass around mmr. We should already have mmrRatio.

File: packages/zoe/src/contracts/loan/updateDebt.js [hunk 6]
```diff
@@ -19,11 +19,8 @@ const BASIS_POINT_DENOMINATOR = 10000n;
  * interestRate (in basis points) is 5 = 5/10,000
  * interest charged this period is 20 loan brand
  */
-export const calculateInterest = (oldDebtValue, interestRate) =>
-  natSafeMath.floorDivide(
-    natSafeMath.multiply(oldDebtValue, interestRate),
-    BASIS_POINT_DENOMINATOR,
-  );
+export const calculateInterest = (oldDebt, interestRate) =>
+  multiplyBy(oldDebt, interestRate);
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/updateDebt.js:23]: Nice!

File: packages/zoe/src/contracts/loan/updateDebt.js [hunk 7]
```diff
@@ -38,6 +35,11 @@ export const makeDebtCalculator = debtCalculatorConfig => {
     configMinusGetDebt,
   } = debtCalculatorConfig;
   let debt = originalDebt;
+  const interestRatio = makeRatio(
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/updateDebt.js:38]: Huh, I'm thinking that `interestRate` in the terms should be `interestRatio` instead and this should be moved outside the contract. What do you think?

**Review by katelynsills — CHANGES_REQUESTED** [review]
callSpread looks good. For the loan, I don't think we need to support mmr as a percent. Is there a reason we want to? We don't know of anything using it and it's not a core contract.

File: packages/zoe/src/contracts/loan/updateDebt.js (hunk 7 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/loan/updateDebt.js:38]: Ok that sounds fine to me

File: packages/zoe/src/contracts/loan/index.js [hunk 8]
```diff
@@ -27,8 +26,8 @@ import { makePercent } from '../../contractSupport/percentMath';
  * loaned amount and interest must be of the same (separate) brand.
  *
  * Terms:
- *  * mmr (default = 150) - the Maintenance Margin Requirement, in
- *    percent. The default is 150, meaning that collateral should be
+ *  * mmr (default = 150) - the Maintenance Margin Requirement, as a
+ *    ratio. The default is 1.5, meaning that collateral should be
```
Comment (katelynsills) [inline packages/zoe/src/contracts/loan/index.js]: We say 150 in the first line and 1.5 in the second. Should it be 150/100 instead for both?

File: packages/zoe/src/contracts/loan/updateDebt.js (hunk 7 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/loan/updateDebt.js:38]: @Chris-Hibbert were you planning on still moving `interestRatio = makeRatio(` out of the contract? If so, should I come back and continue my review then? Everything else looks good to me.

**Review by katelynsills — COMMENTED** [review]
Looks good to me so far, but I wasn't sure if making `interestRate` a ratio was going to be part of this PR or not. If so, I will come back for the final review when that is in. If not, just let me know. Otherwise looks great!

File: packages/zoe/src/contracts/loan/updateDebt.js (hunk 7 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/loan/updateDebt.js:38]: This is now fixed.

File: packages/zoe/src/contracts/loan/index.js (hunk 8 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/loan/index.js]: fixed.

**Review by Chris-Hibbert — COMMENTED** [review]
The current state is that interim values are being seen in `test-ratio.js` for the output expected in error messages about mis-matched brands. When we have a new SES release, these will show the actual brand in both stack traces while debugging and in the messages that the tests can see.  

I want the PR to pass tests because I have a number of other PRs stacked on top.  I'm ambivalent as to whether it would be better to approve this as-is and merge it, then come back later and fix these tests, or leave this branch hanging until the SES fixes are in.

I don't want to make the SES merge harder. I've marked the tests as `test.failing()`, though I wonder if `test.skip()` would be less work for the SES merge.

**Comment by katelynsills** [comment]
What about a compromise - have some tests that don't depend on any brand `allegedName` being logged, then have other tests that confirm that the brand `allegedName` is logged, and mark those as `failing` until the SES stuff is in?

**Comment by Chris-Hibbert** [comment]
> What about a compromise - have some tests that don't depend on any brand allegedName being logged, then have other tests that confirm that the brand allegedName is logged, and mark those as failing until the SES stuff is in?

Excellent idea. That's what I've done. PTAL when you have a moment.

**Review by katelynsills — APPROVED** [review]
LGTM!

**Pull request merged**